azkaban-common/src/main/java/azkaban/executor/container/ContainerCleanupManager.java

```diff
@@ -194,17 +221,70 @@ public void cleanUpContainersInTerminalStatuses() {
             unFinishedStatusesMap);
         activeFlows.addAll(flows.stream().map(ExecutableFlow::getExecutionId).collect(Collectors.toSet()));
       } catch (final ExecutorManagerException e) {
-        logger.error("Unable to obtain current flows executions of status {}", status, e);
+        logger.error("Unable to obtain current flows executions of status {}, cannot cross-check "
+                + "to-be-killed pods, returning", status, e);
+        return result;
       }
     }
-    // Delete the pod if its execution status is finished
+
+    // cross-check: the pod needs to be cleaned up if its execution status is finished
     for (int executionId : currentNameSpacedPods) {
       if (!activeFlows.contains(executionId)) {
-        logger.info("Cleaning up the stale pod and service for finished execution: {}",
-            executionId);
-        deleteContainerQuietly(executionId);
+        result.add(executionId);
       }
     }
+    return result;
+  }
+
+  /**
+   * Get the flows of EXECUTION_STOPPED status
+   */
+  @NotNull
+  Set<Integer> getExecutionStoppedFlows() {
+    Set<Integer> executionStoppedFlows = new HashSet<>();
+    if (this.executionStoppedMap.containsKey(Status.EXECUTION_STOPPED)) {
+      try {
+        List<ExecutableFlow> flows = this.executorLoader.fetchFreshFlowsForStatus(
+            Status.EXECUTION_STOPPED, this.executionStoppedMap);
+        executionStoppedFlows.addAll(
+            flows.stream().map(ExecutableFlow::getExecutionId).collect(Collectors.toSet()));
+      } catch (final ExecutorManagerException e) {
+        logger.error("Unable to obtain current flows executions of Status.EXECUTION_STOPPED", e);
+      }
+    }
+    return executionStoppedFlows;
+  }
+
+  /**
+   * Find the executions that their yarn applications needs to be killed, then find the set of to be
+   * killed yarn applications, and kill them.
+   */
+  public void cleanUpDanglingYarnApplications() {
+    /*
+      1. find those executions of status EXECUTION_STOPPED status within a recent period
+      2. find those executions being killed but the pod is still unfinished, and delete the pods
+      3. for each of the yarn clusters, do:
+        i. get all unfinished yarn applications
+        ii. filter and get the yarn applications that within the union set of above executions
+        iii. kill these yarn applications
+     */
+    logger.info("cleanUpDanglingYarnApplications start ");
+
+    Set<Integer> executionStoppedFlows = getExecutionStoppedFlows();
+    logger.info("Get executionStoppedFlows: " +
+        executionStoppedFlows.stream().map(Object::toString).collect(Collectors.joining(",")));
+
+    // get those flows terminated but containers are still alive (failed to properly killed)
+    Set<Integer> toBeCleanedContainers = getContainersOfTerminatedFlows();
```

Comment:
This part of finding `flows terminated but containers are still alive` is originally used in the other schedule method `cleanUpContainersInTerminalStatuses`.
Theoretically whenever we found these `containers mistakenly still alive` we should kill these containers, while in this PR I am focusing on cleaning up the yarn apps, and don't want to interfere with others.
We can circle back to discuss whether can improve this part later.